gapic/samplegen/samplegen.py

```diff
@@ -697,29 +695,23 @@ def generate_sample(sample,
         )
 
     calling_form = types.CallingForm.method_default(rpc)
+    sample["id"] = (sample["id"] if id_is_unique else
+                    f'{sample["id"]}_{calling_form}')
```

Comment:
Does the samplegen user guide have guidelines for such a situation?
Instead of second guessing the user, it makes me want to say 'screw it, all ids are concatenated with hashes for guaranteed uniqueness'.

Reply:
Or better yet, raise an exception if an ID/calling_form is duplicated.

Reply:
No guidelines, I believe. I think what @hzyi-google has been doing in the monolith is uniquifying with just the calling form first, and then appending a disambiguating number if necessary, but we don't have a requirement that all generators do it the same way.

Reply:
Mandatory hash it is then:
`{id or region tag}_{optional calling form_}{8 digits of hash}.py`
or, if there is no id or region tag,
`{8 digits of hash}.py`

unless have strong opinions against that.

Reply:
IMO it's OK to not allow two samples with the same _user specified_ ID (unless they are empty strings, in which case users chose not to provide and ID)

Reply:
I'm leaning towards unconditional hashing for disambiguation, and not just because it makes my life easier :P
The following was a rough flow of my thought process:
* Simple rules are usually better than complex rules, both for users and developers.
* Smaller sets of rules are usually simpler than larger sets.
* Strict rules are usually simpler than permissive rules.
* Consistent rules are _always_ simpler than rules with caveats, exceptions, explicit corner cases, and hierarchies.

The rules for sample ID, taken from the guide plus some creative interpretation in light of the above, become:
* Start with the "id" tag. If there isn't one, fall back to "region_tag". If there isn't one, generate a random, probably unique ID.
* If the ID is not unique, generate a probably unique disambiguating suffix.

The end result of which is that if the sample author wants helpful, descriptive, unique IDs the onus is on them. The generator is not smarter than the sample author in this regard.

Reply:
I completely agree with your strategy.